hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/snapshot/SnapshotManager.java

```diff
@@ -288,6 +292,19 @@ public INodeDirectory getSnapshottableAncestorDir(final INodesInPath iip)
     }
   }
 
+  public boolean isDescendantOfSnapshotRoot(INodeDirectory dir) {
+    if (dir.isSnapshottable()) {
+      return true;
+    } else {
+      for (INodeDirectory p = dir; p != null; p = p.getParent()) {
```

Comment:
use INode#isAncestorDirectory() instead?

Reply:
I see `SnapshotManager#getSnapshottableAncestorDir` uses INode#isAncestorDirectory(), as `SnapshotManager#snapshottables` is a map, say it has M elements, and the dir path got N elements, my solution is O(log(M) * N) compare to your O(M*N).

I think `SnapshotManager#getSnapshottableAncestorDir` can be optimized, but not in this patch. What do you think?